torch/csrc/jit/passes/peephole_non_tensor.cpp

```diff
@@ -0,0 +1,134 @@
+              constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+          // if an if node's output equals its condition replace output with
+          // condition
+          if (true_val && !false_val) {
+            GRAPH_UPDATE(
+                "Replacing ",
+                n.outputs().at(i)->debugName(),
+                " (True or False) with ",
+                n.cond()->debugName());
+            n.outputs().at(i)->replaceAllUsesWith(n.cond());
+            changed = true;
+          }
+        }
+      } else if (
+          node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+        // if we are comparing a None value with a value that can't be None
+        // replace the output with true if node is __isnot__ or false if node is
+        // __is__
+        AT_ASSERT(node->inputs().size() == 2);
+        for (size_t check_none_index : {0, 1}) {
+          bool input_must_be_none =
+              node->inputs().at(check_none_index)->mustBeNone();
+          bool other_must_not_be_none =
+              node->inputs().at(1 - check_none_index)->mustNotBeNone();
+          if (input_must_be_none && other_must_not_be_none) {
+            WithInsertPoint guard(node);
+            auto output = node->owningGraph()->insertConstant(
+                node->kind() == aten::__isnot__);
+            GRAPH_UPDATE(
+                "Folding ", getHeader(node), " to ", output->debugName());
+            node->output()->replaceAllUsesWith(output);
+            changed = true;
+          }
+        }
+      } else if (
+          node->kind() == prim::unchecked_unwrap_optional ||
+          node->kind() == aten::_unwrap_optional) {
+        // we are unwrapping an input that can't be None, remove the unwrap
+        auto input = node->input();
+        if (input->mustNotBeNone()) {
+          GRAPH_UPDATE(
+              "Unwrapping ",
+              getHeader(node),
+              " as ",
+              node->input(),
+              " can't be optional");
+          node->output()->replaceAllUsesWith(node->input());
+          changed = true;
+        }
+      } else if (node->kind() == prim::unchecked_cast) {
+        // unchecked_cast is not generated for tensor properties, so we are not
+        // losing anything by calling unshapedType here
+        auto input_type = unshapedType(node->input()->type());
+        auto output_type = unshapedType(node->output()->type());
+        if (input_type->isSubtypeOf(output_type)) {
+          GRAPH_UPDATE(
+              "Removing ",
+              getHeader(node),
+              " as input type subtypes output type");
+          node->output()->replaceAllUsesWith(node->input());
+          changed = true;
+        }
+      }
+    }
+    return changed;
+  }
+
+ private:
+  std::shared_ptr<Graph> graph_;
+};
+
+bool PeepholeOptimizeNonTensor(const std::shared_ptr<Graph>& graph) {
+  PeepholeOptimizeNonTensorImpl peephole(graph);
+  bool changed = peephole.run();
```

Comment:
Do we need to apply EliminateDeadCode here like in https://github.com/pytorch/pytorch/blob/72c3ee073f982c69d532435972a16bd9958413bd/torch/csrc/jit/passes/peephole.cpp#L551 ?

Reply:
This is being invoked as part of PeepholeOptimize which has it's own DCE call so i'm not invoking it here